[BCn] TRANSLATION_FAULT in s3tc/bc6/bc7.comp due to out-of-bounds not guarded by robustBuffer

[wrapper_log_2025_07_31_02_17_20..29004.txt](https://github.com/user-attachments/files/21519359/wrapper_log_2025_07_31_02_17_20.29004.txt)
[wrapper_log_2025_07_31_02_07_02.MiSideFull.exe.27097.txt](https://github.com/user-attachments/files/21519358/wrapper_log_2025_07_31_02_07_02.MiSideFull.exe.27097.txt)

MiSide: only bcn 133 and 137 so far, black screens but sound plays

```
[ERROR] Call to DeviceWaitIdle with (device: 0xb40000749e035030) failed with result: -4      (wrapper_tramp_DeviceWaitIdle:3481)
[ERROR] FATAL: vkDeviceWaitIdle(base->dispatch_handle)  failed with a GPU fault, the game will now crash      (wrapper_log_device_fault:184)
[ERROR] Fault counts: addressInfoCount=1, vendorInfoCount=1, vendorBinarySize=0      (wrapper_log_device_fault:209)
[ERROR] + deviceFaultVendorBinary feature not available, cannot dump vendor specific fault info      (wrapper_log_device_fault:221)
[ERROR] --- VULKAN DEVICE FAULT DETECTED ---      (wrapper_log_device_fault:248)
[ERROR] Description: Exception type 0xc3: TRANSLATION_FAULT at level 3. VA: 0x5fffc19020. Access type 0x2: Read. Source id 0x240: Shader core 1. Group 0.      (wrapper_log_device_fault:249)
[ERROR] .pAddressInfos[0]      (wrapper_log_device_fault:251)
  .addressType: VkDeviceFaultAddressTypeEXT = 0x1
  .reportedAddress: VkDeviceAddress = 0x5fffc19020
  .addressPrecision: VkDeviceSize = 0x1000
[ERROR] .pVendorInfos[0]      (wrapper_log_device_fault:255)
  .description: char = 'Exception type 0xc3: TRANSLATION_FAULT at level 3. VA: 0x5fffc19020. Access type 0x2: Read. Source id 0x240: Shader core 1. Group 0.'
  .vendorFaultCode: uint64_t = 0xc3
  .vendorFaultData: uint64_t = 0x1024002c3
[ERROR] --- END FAULT INFO ---      (wrapper_log_device_fault:262)
```

Outlast: black screens but sound plays

```
[ERROR] --- VULKAN DEVICE FAULT DETECTED ---      (wrapper_log_device_fault:248)
[ERROR] Description: Exception type 0xc3: TRANSLATION_FAULT at level 3. VA: 0x5fffc4e020. Access type 0x2: Read. Source id 0x40: Shader core 0. Group 0.      (wrapper_log_device_fault:249)
[ERROR] .pAddressInfos[0]      (wrapper_log_device_fault:251)
  .addressType: VkDeviceFaultAddressTypeEXT = 0x1
  .reportedAddress: VkDeviceAddress = 0x5fffc4e020
  .addressPrecision: VkDeviceSize = 0x1000
[ERROR] .pVendorInfos[0]      (wrapper_log_device_fault:255)
  .description: char = 'Exception type 0xc3: TRANSLATION_FAULT at level 3. VA: 0x5fffc4e020. Access type 0x2: Read. Source id 0x40: Shader core 0. Group 0.'
  .vendorFaultCode: uint64_t = 0xc3
  .vendorFaultData: uint64_t = 0x1004002c3
[ERROR] --- END FAULT INFO ---      (wrapper_log_device_fault:262)
```

```
CreatePipelineLayout (id=25)
  in: device: VkDevice (handle) = 0xb400007e783f88c0 (id=25)
  in: pCreateInfo: VkPipelineLayoutCreateInfo* (id=25)
    .flags: VkPipelineLayoutCreateFlags = 0x0
    .setLayoutCount: uint32_t = 0x0
    .pushConstantRangeCount: uint32_t = 0x0
  in: pAllocator: VkAllocationCallbacks* (id=25)
  out: result: VkResult = 0 (id=25)
  out: *pPipelineLayout: VkPipelineLayout = b400007ee83f0710 (id=25)
CreateShaderModule (id=26)
  in: device: VkDevice (handle) = 0xb400007e783f88c0 (id=26)
  in: pCreateInfo: VkShaderModuleCreateInfo* (id=26)
    .flags: VkShaderModuleCreateFlags = 0x0
    .codeSize: size_t = 0x128
    .pCode: uint32_t* = 0x7c86d2813c
      .spirv(0x7c86d2813c)[0000]: 07230203 00010000 0008000b 00000010 00000000 00020011 00000001 0006000b
      .spirv(0x7c86d2813c)[0020]: 00000001 4c534c47 6474732e 3035342e 00000000 0003000e 00000000 00000001
      .spirv(0x7c86d2813c)[0040]: 0005000f 00000005 00000004 6e69616d 00000000 00060010 00000004 00000011
      .spirv(0x7c86d2813c)[0060]: 00000008 00000008 00000001 00030003 00000002 000001c2 00040005 00000004
      .spirv(0x7c86d2813c)[0080]: 6e69616d 00000000 00040047 0000000f 0000000b 00000019 00020013 00000002
      .spirv(0x7c86d2813c)[00a0]: 00030021 00000003 00000002 00040015 0000000b 00000020 00000000 00040017
      .spirv(0x7c86d2813c)[00c0]: 0000000c 0000000b 00000003 0004002b 0000000b 0000000d 00000008 0004002b
      .spirv(0x7c86d2813c)[00e0]: 0000000b 0000000e 00000001 0006002c 0000000c 0000000f 0000000d 0000000d
      .spirv(0x7c86d2813c)[0100]: 0000000e 00050036 00000002 00000004 00000000 00000003 000200f8 00000005
      .spirv(0x7c86d2813c)[0120]: 000100fd 00010038 
  in: pAllocator: VkAllocationCallbacks* (id=26)
  out: result: VkResult = 0 (id=26)
  out: *pShaderModule: VkShaderModule = b400007d483f2fa0 (id=26)
CreateComputePipelines (id=27)
  in: device: VkDevice (handle) = 0xb400007e783f88c0 (id=27)
  in: pipelineCache: VkPipelineCache = 0 (id=27)
  in: createInfoCount: uint32_t = 1 (id=27)
  in: pCreateInfos[]: VkComputePipelineCreateInfo (id=27)
    pCreateInfos[0]: VkComputePipelineCreateInfo (id=27)
      .flags: VkPipelineCreateFlags = 0x0
      .stage: VkPipelineShaderStageCreateInfo
        .flags: VkPipelineShaderStageCreateFlags = 0x0
        .stage: VkShaderStageFlagBits = 0x20
        .module: VkShaderModule (handle) = 0xb400007d483f2fa0
        .pName: char* = 'main'
      .layout: VkPipelineLayout (handle) = 0xb400007ee83f0710
      .basePipelineHandle: VkPipeline (handle) = 0x0
      .basePipelineIndex: int32_t = 0x0
  in: pAllocator: VkAllocationCallbacks* (id=27)
  out: result: VkResult = 0 (id=27)
  out: pPipelines[]: VkPipeline = 0x7c86fe8f90 (id=27)
    pPipelines[0]: VkPipeline = b400007ee83f3420 (id=27)

...

// Second time
wrapper_CreateDevice: Application requested the following features
Turning on VK_EXT_device_fault for better GPU fault reporting.
Faking extension support for EXT_transform_feedback->geometryStreams
Faking feature support for textureCompressionBC
Faking feature support for multiViewport
Faking feature support for fillModeNonSolid
Faking feature support for shaderClipDistance
Faking feature support for shaderCullDistance
Faking feature support for dualSrcBlend
Faking feature support for vertexPipelineStoresAndAtomics
CreatePipelineLayout (id=1696)
  in: device: VkDevice (handle) = 0xb400007e783fe3e0 (id=1696)
  in: pCreateInfo: VkPipelineLayoutCreateInfo* (id=1696)
    .flags: VkPipelineLayoutCreateFlags = 0x0
    .setLayoutCount: uint32_t = 0x0
    .pushConstantRangeCount: uint32_t = 0x0
  in: pAllocator: VkAllocationCallbacks* (id=1696)
  out: result: VkResult = 0 (id=1696)
  out: *pPipelineLayout: VkPipelineLayout = b400007ee83f8e40 (id=1696)
CreateShaderModule (id=1697)
  in: device: VkDevice (handle) = 0xb400007e783fe3e0 (id=1697)
  in: pCreateInfo: VkShaderModuleCreateInfo* (id=1697)
    .flags: VkShaderModuleCreateFlags = 0x0
    .codeSize: size_t = 0x128
    .pCode: uint32_t* = 0x7c86d2813c
      .spirv(0x7c86d2813c)[0000]: 07230203 00010000 0008000b 00000010 00000000 00020011 00000001 0006000b
      .spirv(0x7c86d2813c)[0020]: 00000001 4c534c47 6474732e 3035342e 00000000 0003000e 00000000 00000001
      .spirv(0x7c86d2813c)[0040]: 0005000f 00000005 00000004 6e69616d 00000000 00060010 00000004 00000011
      .spirv(0x7c86d2813c)[0060]: 00000008 00000008 00000001 00030003 00000002 000001c2 00040005 00000004
      .spirv(0x7c86d2813c)[0080]: 6e69616d 00000000 00040047 0000000f 0000000b 00000019 00020013 00000002
      .spirv(0x7c86d2813c)[00a0]: 00030021 00000003 00000002 00040015 0000000b 00000020 00000000 00040017
      .spirv(0x7c86d2813c)[00c0]: 0000000c 0000000b 00000003 0004002b 0000000b 0000000d 00000008 0004002b
      .spirv(0x7c86d2813c)[00e0]: 0000000b 0000000e 00000001 0006002c 0000000c 0000000f 0000000d 0000000d
      .spirv(0x7c86d2813c)[0100]: 0000000e 00050036 00000002 00000004 00000000 00000003 000200f8 00000005
      .spirv(0x7c86d2813c)[0120]: 000100fd 00010038 
  in: pAllocator: VkAllocationCallbacks* (id=1697)
  out: result: VkResult = 0 (id=1697)
  out: *pShaderModule: VkShaderModule = b400007d483fc460 (id=1697)
CreateComputePipelines (id=1698)
  in: device: VkDevice (handle) = 0xb400007e783fe3e0 (id=1698)
  in: pipelineCache: VkPipelineCache = 0 (id=1698)
  in: createInfoCount: uint32_t = 1 (id=1698)
  in: pCreateInfos[]: VkComputePipelineCreateInfo (id=1698)
    pCreateInfos[0]: VkComputePipelineCreateInfo (id=1698)
      .flags: VkPipelineCreateFlags = 0x0
      .stage: VkPipelineShaderStageCreateInfo
        .flags: VkPipelineShaderStageCreateFlags = 0x0
        .stage: VkShaderStageFlagBits = 0x20
        .module: VkShaderModule (handle) = 0xb400007d483fc460
        .pName: char* = 'main'
      .layout: VkPipelineLayout (handle) = 0xb400007ee83f8e40
      .basePipelineHandle: VkPipeline (handle) = 0x0
      .basePipelineIndex: int32_t = 0x0
  in: pAllocator: VkAllocationCallbacks* (id=1698)
  out: result: VkResult = 0 (id=1698)
  out: pPipelines[]: VkPipeline = 0x7c86fe8f90 (id=1698)
    pPipelines[0]: VkPipeline = b400007ee83f94b0 (id=1698)

...

Emulating support for format=137, count=1
Enabling experimental compute shader mode
  + use_compute_shader=1, use_image_view=1
Submitting one-time commands...
AllocateCommandBuffers (id=2131)
  in: device: VkDevice (handle) = 0xb400007e783f88c0 (id=2131)
  out: result: VkResult = 0 (id=2131)
BeginCommandBuffer (id=2132)
  in: commandBuffer: VkCommandBuffer (handle) = 0xb400007e38448650 (id=2132)
  in: pBeginInfo: VkCommandBufferBeginInfo* (id=2132)
    .flags: VkCommandBufferUsageFlags = 0x1
  out: result: VkResult = 0 (id=2132)
CmdComputeShaderForDecompression: srcBuffer = 0xb400007e683fb530, dstImage = 0xb400007ef8430e50
CmdPipelineBarrier (id=2133)
  in: commandBuffer: VkCommandBuffer (handle) = 0xb400007e38448650 (id=2133)
  in: srcStageMask: VkPipelineStageFlags = 10000 (id=2133)
  in: dstStageMask: VkPipelineStageFlags = 800 (id=2133)
  in: dependencyFlags: VkDependencyFlags = 0 (id=2133)
  in: memoryBarrierCount: uint32_t = 0 (id=2133)
  in: pMemoryBarriers[]: VkMemoryBarrier (id=2133)
  in: bufferMemoryBarrierCount: uint32_t = 1 (id=2133)
  in: pBufferMemoryBarriers[]: VkBufferMemoryBarrier (id=2133)
    pBufferMemoryBarriers[0]: VkBufferMemoryBarrier (id=2133)
      .srcAccessMask: VkAccessFlags = 0x15000
      .dstAccessMask: VkAccessFlags = 0x20
      .srcQueueFamilyIndex: uint32_t = 0xffffffff
      .dstQueueFamilyIndex: uint32_t = 0xffffffff
      .buffer: VkBuffer (handle) = 0xb400007e683fb530
      .offset: VkDeviceSize = 0x5600
      .size: VkDeviceSize = 0xffffffffffffffff
  in: imageMemoryBarrierCount: uint32_t = 1 (id=2133)
  in: pImageMemoryBarriers[]: VkImageMemoryBarrier (id=2133)
    pImageMemoryBarriers[0]: VkImageMemoryBarrier (id=2133)
      .srcAccessMask: VkAccessFlags = 0x0
      .dstAccessMask: VkAccessFlags = 0x40
      .oldLayout: VkImageLayout = 0x7
      .newLayout: VkImageLayout = 0x1
      .srcQueueFamilyIndex: uint32_t = 0xffffffff
      .dstQueueFamilyIndex: uint32_t = 0xffffffff
      .image: VkImage (handle) = 0xb400007ef8430e50
      .subresourceRange: VkImageSubresourceRange
        .aspectMask: VkImageAspectFlags = 0x1
        .baseMipLevel: uint32_t = 0x0
        .levelCount: uint32_t = 0x1
        .baseArrayLayer: uint32_t = 0x0
        .layerCount: uint32_t = 0x1
CreateDescriptorPool (id=2134)
  in: device: VkDevice (handle) = 0xb400007e783f88c0 (id=2134)
  in: pCreateInfo: VkDescriptorPoolCreateInfo* (id=2134)
    .flags: VkDescriptorPoolCreateFlags = 0x0
    .maxSets: uint32_t = 0x1
    .poolSizeCount: uint32_t = 0x3
    .pPoolSizes[0]: VkDescriptorPoolSize*
      .type: VkDescriptorType = 0x7
      .descriptorCount: uint32_t = 0x1
    .pPoolSizes[1]: VkDescriptorPoolSize*
      .type: VkDescriptorType = 0x3
      .descriptorCount: uint32_t = 0x1
    .pPoolSizes[2]: VkDescriptorPoolSize*
      .type: VkDescriptorType = 0x6
      .descriptorCount: uint32_t = 0x1
  in: pAllocator: VkAllocationCallbacks* (id=2134)
  out: result: VkResult = 0 (id=2134)
  out: *pDescriptorPool: VkDescriptorPool = b400007ee8402250 (id=2134)
FlushMappedMemoryRanges (id=2135)
  in: device: VkDevice (handle) = 0xb400007e783f88c0 (id=2135)
  in: memoryRangeCount: uint32_t = 1 (id=2135)
  in: pMemoryRanges[]: VkMappedMemoryRange (id=2135)
    pMemoryRanges[0]: VkMappedMemoryRange (id=2135)
      .memory: VkDeviceMemory (handle) = 0xb400007f083f2d90
      .offset: VkDeviceSize = 0x0
      .size: VkDeviceSize = 0xffffffffffffffff
  out: result: VkResult = 0 (id=2135)
CreateImageView (id=2136)
  in: device: VkDevice (handle) = 0xb400007e783f88c0 (id=2136)
  in: pCreateInfo: VkImageViewCreateInfo* (id=2136)
    .flags: VkImageViewCreateFlags = 0x0
    .image: VkImage (handle) = 0xb400007ef8430e50
    .viewType: VkImageViewType = 0x1
    .format: VkFormat = 0x25
    .components: VkComponentMapping
      .r: VkComponentSwizzle = 0x0
      .g: VkComponentSwizzle = 0x0
      .b: VkComponentSwizzle = 0x0
      .a: VkComponentSwizzle = 0x0
    .subresourceRange: VkImageSubresourceRange
      .aspectMask: VkImageAspectFlags = 0x1
      .baseMipLevel: uint32_t = 0x0
      .levelCount: uint32_t = 0x1
      .baseArrayLayer: uint32_t = 0x0
      .layerCount: uint32_t = 0x1
  in: pAllocator: VkAllocationCallbacks* (id=2136)
  out: result: VkResult = 0 (id=2136)
  out: *pView: VkImageView = b400007ef840d590 (id=2136)
Tracking stagingImageView (0xb400007ef840d590) for 47
CmdBindPipeline (id=2137)
  in: commandBuffer: VkCommandBuffer (handle) = 0xb400007e38448650 (id=2137)
  in: pipelineBindPoint: VkPipelineBindPoint = 1 (id=2137)
  in: pipeline: VkPipeline = b400007ee83f94b0 (id=2137)
Calling CmdDispatch for 0
CmdDispatch (id=2138)
  in: commandBuffer: VkCommandBuffer (handle) = 0xb400007e38448650 (id=2138)
  in: groupCountX: uint32_t = 100 (id=2138)
  in: groupCountY: uint32_t = 80 (id=2138)
  in: groupCountZ: uint32_t = 1 (id=2138)
CmdDispatch Finished for 0
CmdPipelineBarrier (id=2139)
  in: commandBuffer: VkCommandBuffer (handle) = 0xb400007e38448650 (id=2139)
  in: srcStageMask: VkPipelineStageFlags = 800 (id=2139)
  in: dstStageMask: VkPipelineStageFlags = 2000 (id=2139)
  in: dependencyFlags: VkDependencyFlags = 0 (id=2139)
  in: memoryBarrierCount: uint32_t = 0 (id=2139)
  in: pMemoryBarriers[]: VkMemoryBarrier (id=2139)
  in: bufferMemoryBarrierCount: uint32_t = 0 (id=2139)
  in: pBufferMemoryBarriers[]: VkBufferMemoryBarrier (id=2139)
  in: imageMemoryBarrierCount: uint32_t = 1 (id=2139)
  in: pImageMemoryBarriers[]: VkImageMemoryBarrier (id=2139)
    pImageMemoryBarriers[0]: VkImageMemoryBarrier (id=2139)
      .srcAccessMask: VkAccessFlags = 0x40
      .dstAccessMask: VkAccessFlags = 0x820
      .oldLayout: VkImageLayout = 0x1
      .newLayout: VkImageLayout = 0x7
      .srcQueueFamilyIndex: uint32_t = 0xffffffff
      .dstQueueFamilyIndex: uint32_t = 0xffffffff
      .image: VkImage (handle) = 0xb400007ef8430e50
      .subresourceRange: VkImageSubresourceRange
        .aspectMask: VkImageAspectFlags = 0x1
        .baseMipLevel: uint32_t = 0x0
        .levelCount: uint32_t = 0x1
        .baseArrayLayer: uint32_t = 0x0
        .layerCount: uint32_t = 0x1
EndCommandBuffer (id=2140)
  in: commandBuffer: VkCommandBuffer (handle) = 0xb400007e38448650 (id=2140)
  out: result: VkResult = 0 (id=2140)
CreateFence (id=2141)
  in: device: VkDevice (handle) = 0xb400007e783f88c0 (id=2141)
  in: pCreateInfo: VkFenceCreateInfo* (id=2141)
    .flags: VkFenceCreateFlags = 0x0
  in: pAllocator: VkAllocationCallbacks* (id=2141)
  out: result: VkResult = 0 (id=2141)
  out: *pFence: VkFence = b400007ef8435950 (id=2141)
Submitting command buffer to queue 0xb400007dd83e7760
QueueSubmit (id=2142)
  in: queue: VkQueue (handle) = 0xb400007dd83e7760 (id=2142)
  in: submitCount: uint32_t = 1 (id=2142)
  in: pSubmits[]: VkSubmitInfo (id=2142)
    pSubmits[0]: VkSubmitInfo (id=2142)
      .waitSemaphoreCount: uint32_t = 0x0
      .commandBufferCount: uint32_t = 0x1
      .pCommandBuffers[0]: VkCommandBuffer* (handle) = b400007e38448650
      .signalSemaphoreCount: uint32_t = 0x0
  in: fence: VkFence = b400007ef8435950 (id=2142)
  out: result: VkResult = 0 (id=2142)
Waiting for fence 0xb400007ef8435950
WaitForFences (id=2143)
  in: device: VkDevice (handle) = 0xb400007e783f88c0 (id=2143)
  in: fenceCount: uint32_t = 1 (id=2143)
  in: pFences[]: VkFence = 0x7c6e8ce960 (id=2143)
    pFences[0]: VkFence = b400007ef8435950 (id=2143)
  in: waitAll: VkBool32 = 1 (id=2143)
  in: timeout: uint64_t = ffffffffffffffff (id=2143)
  out: result: VkResult = 0 (id=2124)
  out: pPipelines[]: VkPipeline = 0x44daf028 (id=2124)
    pPipelines[0]: VkPipeline = b400007d886ceda0 (id=2124)
DestroyShaderModule (id=2144)
  in: device: VkDevice (handle) = 0xb400007e783f88c0 (id=2144)
  in: shaderModule: VkShaderModule = b400007d483faa20 (id=2144)
  in: pAllocator: VkAllocationCallbacks* (id=2144)
DestroyShaderModule (id=2145)
  in: device: VkDevice (handle) = 0xb400007e783f88c0 (id=2145)
  in: shaderModule: VkShaderModule = b400007d483fad30 (id=2145)
  in: pAllocator: VkAllocationCallbacks* (id=2145)
  out: result: VkResult = -4 (id=2143)
[ERROR] Call to WaitForFences with (device: 0xb400007e2841a040,fenceCount: 1,pFences: 6e8ce960,waitAll: 1,timeout: ffffffff) failed with result: -4
```

Hilariously, the problem is that the cached pipeline was created for a VkDevice associated with a Vulkan layer. On many drivers, these would all coalesce to the same VkDevice, but not on this specific driver.

This is the most plausible hypothesis, validating it now.

--- a/.gitignore
+++ b/.gitignore
@@ -5,3 +5,4 @@
 *.out
 /build
 .venv/
+s3tc2.comp

--- a/src/vulkan/wrapper/wrapper_device.c
+++ b/src/vulkan/wrapper/wrapper_device.c
@@ -43,16 +43,6 @@ static const uint32_t bc7_iv_spv[] = {
 #include "bc7_iv.spv.h"
 };
 
-// A struct to hold the state required by our interceptor
-typedef struct InterceptorState {
-   VkDevice device;
-   VkPipeline pipeline;
-   VkPipelineLayout pipelineLayout;
-   VkDescriptorSetLayout descriptorSetLayout;
-   VkBuffer constantsBuffer;
-   VkDeviceMemory constantsBufferMemory;
-} InterceptorState;
-
 // Initializes the Vulkan objects needed for the compute dispatch.
 // Call this after intercepting vkCreateDevice.
 static VkResult InterceptorState_Init(InterceptorState* state, VkDevice device, size_t spv_size, const uint32_t* spv_code, bool use_image_view, int bc_mode);
@@ -60,10 +50,6 @@ static VkResult InterceptorState_Init(InterceptorState* state, VkDevice device, 
 // Cleans up the Vulkan objects.
 // Call this before intercepting vkDestroyDevice.
 static void InterceptorState_Cleanup(InterceptorState* state);
-
-static InterceptorState g_interceptorState_s3tc = {0};
-static InterceptorState g_interceptorState_bc6 = {0};
-static InterceptorState g_interceptorState_bc7 = {0};
 
 const struct vk_device_extension_table wrapper_device_extensions =
 {
@@ -333,7 +319,7 @@ wrapper_CreateDevice(VkPhysicalDevice physicalDevice,
 
    // Initialize the BCn interceptor states
    bool use_image_view = use_image_view_mode();
-   result = InterceptorState_Init(&g_interceptorState_s3tc, 
+   result = InterceptorState_Init(&device->s3tc, 
       wrapper_device_to_handle(device), 
       use_image_view ? sizeof(s3tc_iv_spv) : sizeof(s3tc_spv), 
       use_image_view ? s3tc_iv_spv : s3tc_spv, 
@@ -342,7 +328,7 @@ wrapper_CreateDevice(VkPhysicalDevice physicalDevice,
       WLOGE("Failed to initialize InterceptorState for s3tc");
       return vk_error(physical_device, result);
    }
-   result = InterceptorState_Init(&g_interceptorState_bc6, 
+   result = InterceptorState_Init(&device->bc6, 
       wrapper_device_to_handle(device), 
       use_image_view ? sizeof(bc6_iv_spv) : sizeof(bc6_spv), 
       use_image_view ? bc6_iv_spv : bc6_spv, 
@@ -351,7 +337,7 @@ wrapper_CreateDevice(VkPhysicalDevice physicalDevice,
       WLOGE("Failed to initialize InterceptorState for bc6");
       return vk_error(physical_device, result);
    }
-   result = InterceptorState_Init(&g_interceptorState_bc7, 
+   result = InterceptorState_Init(&device->bc7, 
       wrapper_device_to_handle(device), 
       use_image_view ? sizeof(bc7_iv_spv) : sizeof(bc7_spv), 
       use_image_view ? bc7_iv_spv : bc7_spv, 
@@ -1798,13 +1784,13 @@ wrapper_CmdCopyBufferToImage(VkCommandBuffer commandBuffer,
       }
 
       if (use_compute_shader) {
-         struct InterceptorState* state = &g_interceptorState_s3tc;
+         struct InterceptorState* state = &_device->s3tc;
          if (wimg->original_format == VK_FORMAT_BC7_UNORM_BLOCK 
             || wimg->original_format == VK_FORMAT_BC7_SRGB_BLOCK) {
-            state = &g_interceptorState_bc7;
+            state = &_device->bc7;
          } else if (wimg->original_format == VK_FORMAT_BC6H_SFLOAT_BLOCK 
             || wimg->original_format == VK_FORMAT_BC6H_UFLOAT_BLOCK) {
-            state = &g_interceptorState_bc6;
+            state = &_device->bc6;
          }
          struct CmdComputeShaderForDecompressionArgs args = {
             ._device = _device,

--- a/src/vulkan/wrapper/wrapper_objects.h
+++ b/src/vulkan/wrapper/wrapper_objects.h
@@ -67,6 +67,16 @@ struct wrapper_queue {
 VK_DEFINE_HANDLE_CASTS(wrapper_queue, vk.base, VkQueue,
                        VK_OBJECT_TYPE_QUEUE)
 
+// A struct to hold the state required by our interceptor
+typedef struct InterceptorState {
+   VkDevice device;
+   VkPipeline pipeline;
+   VkPipelineLayout pipelineLayout;
+   VkDescriptorSetLayout descriptorSetLayout;
+   VkBuffer constantsBuffer;
+   VkDeviceMemory constantsBufferMemory;
+} InterceptorState;
+
 struct wrapper_device {
    struct vk_device vk;
 
@@ -88,6 +98,10 @@ struct wrapper_device {
    uint32_t graphics_queue_idx;
 
    VkCommandPool computePool;
+
+   InterceptorState s3tc;
+   InterceptorState bc6;
+   InterceptorState bc7;
 };
 
 VK_DEFINE_HANDLE_CASTS(wrapper_device, vk.base, VkDevice,

